cmd/internal/exec/std_in_reader.go

```diff
@@ -0,0 +1,47 @@
+package exec
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+)
+
+type stdInReader struct {
+	Reader  io.Reader
+	Timeout time.Duration
+}
+
+func (sr *stdInReader) ReadInput() (string, error) {
+	input := make(chan string, 1)
+	done := make(chan bool, 1)
+	var lines []string
+	in := bufio.NewScanner(sr.Reader)
+	go func(isDone <-chan bool) {
+		for {
+			select {
+			case <-isDone:
+				return
+			default:
+				in.Scan()
+				line := in.Text()
+				if len(line) == 0 {
+					input <- strings.Join(lines, "\n")
+					return
+				}
+				lines = append(lines, line)
+			}
+		}
+	}(done)
+
+	for {
+		select {
+		case i := <-input:
+			return i, nil
+		case <-time.After(sr.Timeout):
+			done <- true
```

Comment:
I think `close(done)` would be more slightly more idiomatic? Closed channels return their type's zero value; but we don't care for that value anyhow.

Reply:
TIL 👍